[Add] Google Map ZoomPanel, 주유소 Marker 구현

shinsewon opened the pull request:
혜원님 데이터 뿌리고 구현하는 과정에서 어쩌다보니 작업양이 많아졌네요.🙄
시간 되실때 확인 부탁드려요 :)
그리고 현재 실행하면 맵에서 위치를 변경될때마다 마커 data가 refetch되는데, 이 부분은 다음번 작업에서 update 할 예정입니다!

1. Zoom Panel Component 구현 및 Zoom 기능 추가
2. useQuery 라이브러리 설치 및 적용
3. 위도 경도 변경 proj4 라이브러리 설치 및 적용
4. api 파일 작성
5. ios http 통신 허용작업
6. 내 현재위치 state 상태 및 marker구현 (이미지 임시 구현)
8. 오피넷 API 5km내 반경 주유소 데이터 구현
9. 기존 GAS 명칭 OIl 명칭으로 변경
10. 농협, 기타 알뜰 주유소 마커 이미지 추가

shinsewon committed:
[Add] Google Map ZoomPanel, 주유소 Marker 구현
1.Zoom Panel Component 구현 및 Zoom 기능 추가
2. useQuery 라이브러리 설치 및 적용
3. 위도 경도 변경 proj4 라이브러리 설치 및 적용
4. api 파일 작성
5. ios http 통신 허용작업
6. 내 현재위치 state 상태 및 marker구현 (이미지 임시 구현)
8. 오피넷 API 5km내 반경 주유소 데이터 구현
9. 기존 GAS 명칭 OIl 명칭으로 변경
10. 농협, 기타 알뜰 주유소 마커 이미지 추가

dohye1 commented on src/api/types.ts:
```diff
@@ -0,0 +1,18 @@
+export type OilStationType = {
+  UNI_ID: string;
+  POLL_DIV_CD: string;
+  OS_NM: string;
+  PRICE: number;
+  DISTANCE: number;
+  GIS_X_COOR: number;
+  GIS_Y_COOR: number;
+};
+
+export type AroundAllOilStationParamsType = {
+  x: number;
+  y: number;
+  radius: number;
+  prodcd: string;
+  sort: number;
+};
+
```
이 파일은 @types/ 파일 하위에 새로 파일을 만들어서 쓰는게 어떨까요?

shinsewon replied:
@type/map 안에 넣을까요? api 타입들도 글로벌 타입에 정리를 해야할지 고민하고 있었는데 그럼 map에 관련된 타입들은 다 @type/map안에 넣을게요 ㅎㅎ

dohye1 replied:
map.d.ts파일이 너무 커진다싶으시면

map.d.ts : 맵과관련된 타입
oil.d.ts : 주유소와 관련된 타입
parking.d.ts : 주차장과 관련된 타입

이렇게 파일을 3개로 나눠도 좋을것같습니다.

dohye1 commented on src/utils/map.ts:
```diff
@@ -0,0 +1,41 @@
+  };
+};
+
+export const geoCurrentLocation = (
+  setState: Dispatch<SetStateAction<LatLng>>,
+) => {
+  Geolocation.getCurrentPosition(
+    position => {
+      const latitude = +JSON.stringify(position.coords.latitude);
+      const longitude = +JSON.stringify(position.coords.longitude);
+
+      setState({latitude, longitude});
+    },
+    error => {
+      console.log(error.code, error.message);
+    },
+    {enableHighAccuracy: true, timeout: 15000, maximumAge: 10000},
+  );
+};
```
setState를 인수로 넘겨서 일반 함수에서쓰는것도 동작은 할지모르지만,
hooks을 사용하는경우엔 custom hook파일을 만들어서 쓰는게 더 맞을것같다는생각이드네용..!

일반적으로 hook은 리액트 내부 로직에의해 돌아가고있으니 리액트의 룰에 최대한 맞춰서 쓰는게 어떨까용..?

shinsewon replied:
오.. 그렇네요. 막상 다른 로직에는 잘 hooks로 만들고 이런건 안만들고 있었네요. 혜원님 덕에 또 깨우침을 얻습니다 ㅎㅎ

dohye1 commented on src/api/apiService.ts:
```diff
@@ -0,0 +1,50 @@
+  AxiosInstance,
+} from "axios";
+
+export class APIService {
```
class로 선언해서 사용하시는 이유가 혹시 무엇인가요?

dohye1 replied:
지금 프로젝트에서 메서드를 함수형으로 만들어서 사용하고있는데 혹시 class를 사용하시는이유가 궁금합니다!

shinsewon replied:
이 부분은 제 지인 cto님한테 배운 내용인데요.

첫번째 이유로는, typescript에서 class는 기존 javascript에는 없는 3가지 접근 제어자가 생깁니다.

* public : 디폴트값, 어디에서나 접근가능.
* protected : 상속받은 하위클래스만 접근가능.
* private : 선언한 클래스 내에서만 접근가능.
* static :  별도 인스턴스를 만들지 않고도 static 함수를 사용하여 인스턴스를 만들 수 있음

이러한 접근제어자들을 적절하게 사용하여, 보다 더 나은 모듈화 및 추상화를 시킬수 있습니다.

두번째 이유로는, 추상적이기 때문입니다. "new(인스턴스화) 하였을 시 어떤 객체가 나오는가?" 에 대한 답이 없어, 여러 api 인스턴스들을 만들때 객체지향 프로그래밍에서 상속관계를 나타내기 위함이라고 배워서 저렇게 적용해서 사용했습니다 :)

dohye1 commented on src/api/map.ts:
```diff
@@ -0,0 +1,19 @@
+import {opinetAPI} from './api';
+import {AroundAllOilStationParamsType, OilStationType} from './types';
```
타입들은 d.ts파일로 따로 만들어서 사용하는게 좋을것같습니다!

shinsewon replied:
네 적용했습니다~~

dohye1 commented on src/components/Map/GasStationMarker.tsx:
```diff
@@ -5,46 +5,53 @@ import {Image} from 'react-native';
-  description?: string;
-  brandName?: string;
+  title: string;
+  brandName: string;
```
brandName : OIL_STATIONS 이렇게 해도되지않나용?

shinsewon replied:
안그래도 놓친 부분들이었는데 이름 다 통일 시킬게요 ㅎㅎ

dohye1 commented on src/components/Map/GoogleMap.tsx:
```diff
@@ -1,36 +1,135 @@
+import {getAroundAllOilStation, OilStationType} from 'api';
+import images from 'assets/images';
+
+//아래 proj4 라이브러리는 google map의 지도 위치 표기 방법은 WGS84방식, 오피넷의 위치 표기방식은 TM128방식이므로, 이를 서로 변경해주는 작업입니다.
+const WGS84 = 'WGS84';
+const TM128 = 'TM128';
+
+proj4.defs('WGS84', '+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs');
+proj4.defs(
+  'TM128',
+  '+proj=tmerc +lat_0=38 +lon_0=128 +k=0.9999 +x_0=400000 +y_0=600000 +ellps=bessel +units=m +no_defs +towgs84=-115.80,474.99,674.11,1.16,-2.31,-1.63,6.43',
+);
```
오 처음보는 방법이에요 신기하네용

dohye1 commented on src/components/Map/GoogleMap.tsx:
```diff
@@ -1,36 +1,135 @@
-            brandName={marker.brandName}
-            coordinate={marker.location}
-            onPress={() => console.log('임시 클릭')}
-          />
-        ))}
-      </MapView>
+      {region && (
+        <>
+          <MapView
+            ref={mapRef}
+            style={styles.map}
```
이렇게도 좋긴한데,

region이 없을때는 spinner를 보여줘도 좋을것같아요

이런식으로?
```

if(!region){
  return <spinner />
}

return <FlewView>
    <MapView>~~~~
```

shinsewon committed:
1. useGetCurrentPosition custom hook 추가
2. api map type -> global type 변경
3. Google Map region 분기처리 로직 변경
4. gas -> oil 명칭 변경
